--- 2mozilla/toolkit/mozapps/installer/packager.orig	2008-08-27 12:59:01.000000000 +0800
+++ 2mozilla/toolkit/mozapps/installer/packager.mk	2008-08-27 12:59:21.000000000 +0800
@@ -314,8 +314,7 @@
 ifdef MOZ_PKG_MANIFEST
 	$(RM) -rf $(DIST)/xpt
 	$(PERL) -I$(topsrcdir)/xpinstall/packager -e 'use Packager; \
-	  Packager::Copy("$(DIST)", "$(DIST)/$(MOZ_PKG_APPNAME)", \
-	                 "$(MOZ_PKG_MANIFEST)", "$(PKGCP_OS)", 1, 0, 1);'
+	  Packager::Copy("$(DIST)", "$(DIST)/$(MOZ_PKG_APPNAME)", "$(MOZ_PKG_MANIFEST)", "$(PKGCP_OS)", 1, 0, 1);'
 	$(PERL) $(topsrcdir)/xpinstall/packager/xptlink.pl -s $(DIST) -d $(DIST)/xpt -f $(DIST)/$(MOZ_PKG_APPNAME)/components -v
 else # !MOZ_PKG_MANIFEST
 ifeq ($(MOZ_PKG_FORMAT),DMG)
