--- mozilla.orig/gfx/src/thebes/nsThebesDeviceContext.cpp.orig	2008-08-29 14:19:02.000000000 +0800
+++ mozilla/gfx/src/thebes/nsThebesDeviceContext.cpp	2008-08-29 14:20:01.000000000 +0800
@@ -174,7 +174,8 @@
 #if defined(MOZ_ENABLE_GTK2)
         GdkScreen *screen = gdk_screen_get_default();
         gtk_settings_get_for_screen(screen); // Make sure init is run so we have a resolution
-        PRInt32 OSVal = PRInt32(round(gdk_screen_get_resolution(screen)));
+        printf("$ @@@@@@@@@@@@@@@@ nsThebesDeviceContext.cpp gdk_screen_get_resolution here @@@@@@@@@ $$$$$$$$$$$$$$ \n");
+        PRInt32 OSVal = PRInt32((int)(gdk_screen_get_resolution(screen)));
 
         if (prefDPI == 0) // Force the use of the OS dpi
             dpi = OSVal;
