diff -r -C3 gcc-4.0.4.orig/libstdc++-v3/config/locale/uclibc/monetary_members.cc gcc-4.0.4/libstdc++-v3/config/locale/uclibc/monetary_members.cc
*** gcc-4.0.4.orig/libstdc++-v3/config/locale/uclibc/monetary_members.cc	2008-08-20 17:59:42.000000000 +0800
--- gcc-4.0.4/libstdc++-v3/config/locale/uclibc/monetary_members.cc	2008-08-21 09:39:29.000000000 +0800
***************
*** 405,411 ****
  # ifdef __UCLIBC_HAS_XLOCALE__
  	  _M_data->_M_decimal_point = __cloc->decimal_point_wc;
  	  _M_data->_M_thousands_sep = __cloc->thousands_sep_wc;
! # else
  	  _M_data->_M_decimal_point = __global_locale->decimal_point_wc;
  	  _M_data->_M_thousands_sep = __global_locale->thousands_sep_wc;
  # endif
--- 405,411 ----
  # ifdef __UCLIBC_HAS_XLOCALE__
  	  _M_data->_M_decimal_point = __cloc->decimal_point_wc;
  	  _M_data->_M_thousands_sep = __cloc->thousands_sep_wc;
! # elif defined __UCLIBC_HAS_LOCALE__
  	  _M_data->_M_decimal_point = __global_locale->decimal_point_wc;
  	  _M_data->_M_thousands_sep = __global_locale->thousands_sep_wc;
  # endif
***************
*** 560,566 ****
  # ifdef __UCLIBC_HAS_XLOCALE__
  	  _M_data->_M_decimal_point = __cloc->decimal_point_wc;
  	  _M_data->_M_thousands_sep = __cloc->thousands_sep_wc;
! # else
  	  _M_data->_M_decimal_point = __global_locale->decimal_point_wc;
  	  _M_data->_M_thousands_sep = __global_locale->thousands_sep_wc;
  # endif
--- 560,566 ----
  # ifdef __UCLIBC_HAS_XLOCALE__
  	  _M_data->_M_decimal_point = __cloc->decimal_point_wc;
  	  _M_data->_M_thousands_sep = __cloc->thousands_sep_wc;
! # elif defined __UCLIBC_HAS_LOCALE__
  	  _M_data->_M_decimal_point = __global_locale->decimal_point_wc;
  	  _M_data->_M_thousands_sep = __global_locale->thousands_sep_wc;
  # endif
diff -r -C3 gcc-4.0.4.orig/libstdc++-v3/config/locale/uclibc/numeric_members.cc gcc-4.0.4/libstdc++-v3/config/locale/uclibc/numeric_members.cc
*** gcc-4.0.4.orig/libstdc++-v3/config/locale/uclibc/numeric_members.cc	2008-08-20 17:59:42.000000000 +0800
--- gcc-4.0.4/libstdc++-v3/config/locale/uclibc/numeric_members.cc	2008-08-21 09:47:59.000000000 +0800
***************
*** 137,143 ****
  # ifdef __UCLIBC_HAS_XLOCALE__
  	  _M_data->_M_decimal_point = __cloc->decimal_point_wc;
  	  _M_data->_M_thousands_sep = __cloc->thousands_sep_wc;
! # else
  	  _M_data->_M_decimal_point = __global_locale->decimal_point_wc;
  	  _M_data->_M_thousands_sep = __global_locale->thousands_sep_wc;
  # endif
--- 137,143 ----
  # ifdef __UCLIBC_HAS_XLOCALE__
  	  _M_data->_M_decimal_point = __cloc->decimal_point_wc;
  	  _M_data->_M_thousands_sep = __cloc->thousands_sep_wc;
! # elif __UCLIBC_HAS_LOCALE__
  	  _M_data->_M_decimal_point = __global_locale->decimal_point_wc;
  	  _M_data->_M_thousands_sep = __global_locale->thousands_sep_wc;
  # endif
